File: src/test/java/com/coxautodev/graphql/tools/RelayConnectionTest.java
package com.coxautodev.graphql.tools;

import graphql.GraphQL;
import graphql.execution.AsyncExecutionStrategy;
import graphql.relay.Connection;
import graphql.relay.SimpleListConnection;
import graphql.schema.DataFetchingEnvironment;
import graphql.schema.GraphQLSchema;
import groovy.lang.Closure;
import org.junit.Test;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

public class RelayConnectionTest {

    @Test
    public void compiles() {
        GraphQLSchema schema = SchemaParser.newParser().file("RelayConnection.graphqls")
                .resolvers(new QueryResolver())
                .dictionary(User.class)
                .build()
                .makeExecutableSchema();

        GraphQL gql = GraphQL.newGraphQL(schema)
                .queryExecutionStrategy(new AsyncExecutionStrategy())
                .build();

        Map<String, Object> variables = new HashMap<>();
        variables.put("limit", 10);
        Utils.assertNoGraphQlErrors(gql, variables, new Closure<String>(null) {
            @Override
            public String call() {
                return "query {\n" +
                        "   users {\n" +
                        "       edges {\n" +
                        "           cursor\n" +
                        "           node {\n" +
                        "               id\n" +
                        "               name\n" +
                        "           }\n" +
                        "       },\n" +
                        "       pageInfo {\n" +
                        "           hasPreviousPage,\n" +
                        "           hasNextPage\n" +
                        "           startCursor\n" +
                        "           endCursor\n" +
                        "       }\n" +
                        "   }\n" +
                        "}";
            }
        });
    }

    static class QueryResolver implements GraphQLQueryResolver {
        public Connection<User> users(int first, String after, DataFetchingEnvironment env) {
            return new SimpleListConnection<>(Collections.singletonList(new User(1L, "Luke"))).get(env);
        }
    }

    static class User {
        Long id;
        String name;

        public User(Long id, String name) {
            this.id = id;
            this.name = name;
        }
    }


}

File: src/test/kotlin/graphql/kickstart/tools/RelayConnectionTest.kt
package graphql.kickstart.tools

import graphql.GraphQL
import graphql.execution.AsyncExecutionStrategy
import graphql.relay.Connection
import graphql.relay.SimpleListConnection
import graphql.schema.DataFetchingEnvironment
import org.junit.Assert
import org.junit.Test

class RelayConnectionTest {

    @Test
    fun `should compile relay schema when not using @connection directive`() {
        val schema = SchemaParser.newParser().schemaString("""
              type Query {
                users(first: Int, after: String): UserConnection
                otherTypes: AnotherTypeConnection
              }
      
              type UserConnection {
                edges: [UserEdge!]!
                pageInfo: PageInfo!
              }
      
              type UserEdge {
                node: User!
              } 
              
              type User {
                id: ID!
                name: String
              }
      
              type PageInfo {
                hasNextPage: Boolean
              }
      
              type AnotherTypeConnection {
                edges: [AnotherTypeEdge!]!
              }
      
              type AnotherTypeEdge {
                node: AnotherType!
              }
      
              type AnotherType {
                echo: String
              }
            """)
            .resolvers(QueryResolver())
            .build()
            .makeExecutableSchema()

        val gql = GraphQL.newGraphQL(schema)
            .queryExecutionStrategy(AsyncExecutionStrategy())
            .build()

        val result = gql.execute("""
          query {
            users {
              edges {
                node {
                  id
                  name
                }
              }
            }
            otherTypes {
              edges {
                node {
                  echo
                }
              }
            }
          }
        """)

        val expected = mapOf(
            "users" to mapOf(
                "edges" to listOf(
                    mapOf("node" to
                        mapOf("id" to "1", "name" to "Luke")
                    )
                )
            ),
            "otherTypes" to mapOf(
                "edges" to listOf(
                    mapOf("node" to
                        mapOf("echo" to "echo-o-o")
                    )
                )
            )
        )

        Assert.assertEquals(expected, result.getData<Map<String, List<*>>>())
    }

    @Test
    fun `should compile relay schema when using @connection directive`() {
        val schema = SchemaParser.newParser()
                .file("RelayConnection.graphqls")
                .resolvers(QueryResolver())
                .dictionary(User::class.java)
                .build()
                .makeExecutableSchema()

        val gql = GraphQL.newGraphQL(schema)
                .queryExecutionStrategy(AsyncExecutionStrategy())
                .build()

        assertNoGraphQlErrors(gql) {
            """
                  query {
                    users {
                      edges {
                        cursor
                        node {
                          id
                          name
                        }
                      }
                      pageInfo {
                        hasPreviousPage
                        hasNextPage
                        startCursor
                        endCursor
                      }
                    }
                  }
            """
        }
    }

    private class QueryResolver : GraphQLQueryResolver {
        fun users(first: Int?, after: String?, env: DataFetchingEnvironment): Connection<User> {
            return SimpleListConnection(listOf(User(1L, "Luke"))).get(env)
        }

        fun otherTypes(env: DataFetchingEnvironment): Connection<AnotherType> {
            return SimpleListConnection(listOf(AnotherType("echo-o-o"))).get(env)
        }
    }

    private data class User(
        val id: Long,
        val name: String
    )

    private data class AnotherType(
        val echo: String
    )
}
